Improve social auth user-facing error handling

mateuszmandera opened the pull request:
Our social auth error handling used to completely not account for ``SOCIAL_AUTH_SUBDOMAIN``, leading to the user getting stuck on the auth subdomain, which could be confusing. I talked more about it in https://chat.zulip.org/#narrow/stream/3-backend/topic/SOCIAL_AUTH_SUBDOMAIN

We fix error-handling by redirecting users to appropriate endpoints on their realm's URL, and then give the subdomain an error page when accessed directly  to inform the user they shouldn't be there:
![auth_subdomain](https://user-images.githubusercontent.com/45007152/74611624-62062180-50f5-11ea-9a4e-e7d55a09907a.png)

Then, building on top of those changes with an altered commit from https://github.com/zulip/zulip/pull/13844 (adjusting for the feedback received too), fixing #13126.
![error](https://user-images.githubusercontent.com/45007152/74611662-ca550300-50f5-11ea-8454-509344c22a9c.png)

timabbott commented:
This looks generally great.  A few requests:
* We should change "is no longer active" to "has been deactivated" -- that seems a lot more clear as to what's going on.
* We should make the `inactive_realm` check before the `inactive_user` check in the social auth code path; it seems like if the realm is deactivated, we shouldn't get as far as to check whether the user is. (Just a quick if statement swap I noticed we should do).
* Can we set `SOCIAL_AUTH_SUBDOMAIN` in `dev_settings.py` so that one can test the new page in the development environment without modifying settings?  
* For the `ROOT_DOMAIN_ALIASES` change, that's probably a good idea.  Two questions: (1) If we forget to update `/etc/zulip/settings.py` to remove `auth` from ROOT_DOMAIN_ALIASES, does something bad happen?   If so we'll want to be careful when deploying this and mention the change in our "Upgrade notes".  And (2) We should make sure you can't register an organization with `SOCIAL_AUTH_SUBDOMAIN` as your subdomain when making a new organization, including having a unit test for that failure mode.

mateuszmandera commented:
Pushed commits with each requested change.

> We should make the inactive_realm check before the inactive_user check in the social auth code path; it seems like if the realm is deactivated, we shouldn't get as far as to check whether the user is. (Just a quick if statement swap I noticed we should do).

Good catch! I think the same should be done in ``is_user_active`` so I swapped the order there too in the commit.

> If we forget to update /etc/zulip/settings.py to remove auth from ROOT_DOMAIN_ALIASES, does something bad happen?

Nothing awful, ``get_subdomain(request)`` will simply return ``''`` for it - so the new error page won't show up. Probably worth adding an entry in Upgrade notes?

mateuszmandera commented:
Fixed a conflict.

timabbott commented on zproject/backends.py:
```diff
@@ -936,15 +936,25 @@ def dict_representation(cls) -> List[ExternalAuthMethodDictT]:
 
-def redirect_deactivated_user_to_login() -> HttpResponseRedirect:
+def redirect_to_signup(realm: Realm) -> HttpResponseRedirect:
+    login_url = reverse('register')
```
This should probably be `signup_url`?

mateuszmandera replied:
Fixed, thanks!

timabbott commented on zproject/backends.py:
```diff
@@ -936,15 +936,25 @@ def dict_representation(cls) -> List[ExternalAuthMethodDictT]:
+    return HttpResponseRedirect(redirect_url)
+
+def redirect_to_login(realm: Realm) -> HttpResponseRedirect:
+    login_url = reverse('zerver.views.auth.login_page', kwargs = {'template_name': 'zerver/login.html'})
```
At some point we should look at whether we can get rid of `/accounts/login` as a URL; it's not user-facing so it seems like we should be able to eliminate it.  But that can be a follow-up.

mateuszmandera replied:
To clarify, you mean getting rid of these
```
    # We have two entries for accounts/login; only the first one is
    # used for URL resolution.  The second here is to allow
    # reverse("django.contrib.auth.views.login") in templates to
    # return `/accounts/login/`.
    url(r'^accounts/login/$', zerver.views.auth.login_page,
        {'template_name': 'zerver/login.html'}, name='zerver.views.auth.login_page'),
    url(r'^accounts/login/$', LoginView.as_view(template_name='zerver/login.html'),
        name='django.contrib.auth.views.login'),
```

and using only ``/login`` for displaying the login_page?

timabbott replied:
Yeah.  I'm not sure where we actually use /accounts/login, but it seems like it should be unnecessary.

mateuszmandera replied:
Yup, I was confused by why we had that endpoint duplicated with /login/. Could be good to remove this as a followup.

timabbott replied:
I think there's probably some ancient history here where we needed `reverse('django.contrib.auth.views.login')` or similar to work, perhaps due to a third-party library.  But I doubt that's still something we need to do, so it'd be nice to clean this up.

chdinesh1089 commented on zerver/views/auth.py:
```diff
@@ -614,13 +617,21 @@ def sort(lst: List[UserProfile]) -> List[UserProfile]:
             pass
 
-    context['deactivated_account_error'] = DEACTIVATED_ACCOUNT_ERROR
+    deactivated_email = request.GET.get('is_deactivated')
+    if deactivated_email is None:
+        return
+    try:
+        validate_email(deactivated_email)
+        context['deactivated_account_error'] = mark_safe(DEACTIVATED_ACCOUNT_ERROR %
+                                                         {'username': deactivated_email})
+    except ValidationError:
+        logging.info("Invalid email in is_deactivated param to login page: %s", deactivated_email)
```
Nothing seems wrong but I'm curious why are we interested in handling this case. I understand that we show the error for something that isn't an email but I don't see any way this could happen unless someone does a GET to `?is_deactivated=<invalid email format>`(which I feel shouldn't be any concern to us) as we send `validated_email` here `return redirect_deactivated_user_to_login(realm, return_data['validated_email'])` in `social_auth_finish`.

mateuszmandera replied:
I'd prefer it not to be for someone to send the user to `/login/?is_deactivate=<evil message>` with `evil_message` including for example `(Please send your login and password to my_evil_email@example.com to regain access)` or something like that, leading to that being displayed inside the error message shown to the user :slightly_smiling_face:

chdinesh1089 replied:
That's an interesting scenario to consider. Thanks!

mateuszmandera commented:
I don't understand why CodeQL is complaining about logging a username :thinking:

chdinesh1089 commented:
> I don't understand why CodeQL is complaining about logging a username 🤔

Apparently it's some issue on GitHub's side. It's failing for all other PRs(on master) too.

timabbott commented:
To what extent does this PR have unmerged work?  We might want to rebase past the black migration if it does.

mateuszmandera commented on zproject/backends.py:
```diff
@@ -1303,11 +1303,23 @@ def dict_representation(cls, realm: Optional[Realm] = None) -> List[ExternalAuth
-def redirect_deactivated_user_to_login() -> HttpResponseRedirect:
+def redirect_to_signup(realm: Realm) -> HttpResponseRedirect:
+    signup_url = reverse("register")
+    redirect_url = realm.uri + signup_url
```
Would we prefer to use `urljoin`s for such operations even on urls fully under our control?

mateuszmandera commented:
@timabbott I think all these commits should be relevant, so rebase. Also added a fix for the thing from https://chat.zulip.org/#narrow/stream/31-production-help/topic/lots.20of.20emails.20w.2F.20error.20messages/near/1212636 at the end

I think the CodeQL error can be dismissed becuase we need to log the email there?

timabbott commented on zerver/views/auth.py:
```diff
@@ -658,13 +661,22 @@ def redirect_to_deactivation_notice() -> HttpResponse:
+            DEACTIVATED_ACCOUNT_ERROR.format(username=deactivated_email)
+        )
+    except ValidationError:
+        logging.info("Invalid email in is_deactivated param to login page: %s", deactivated_email)
 
```
`deactivated_email` is an HTTP GET parameter under the control of someone creating a link; is it safe to render it into the `context` with `mark_safe`, or could this be used as an XSS attack vector?

mateuszmandera replied:
Hmm, this was secured by the `validate_email` call - any XSS attempt should fail because the necessary characters are forbidden in emails - but doesn't hurt to be more careful here so I added an `escape` call.

timabbott commented on zerver/tests/test_signup.py:
```diff
@@ -736,20 +738,21 @@ def test_login_nonexist_user(self) -> None:
-            result,
+        expected_error = (
+            f"Your Zulip account {email} is not a member of the "
+            + "organization associated with this subdomain."
```
I kinda feel like this "wrong subdomain" error category should be removed; it feels like we should just tell the user it was a failed login rather than leaking their credentials were correct but for the wrong domain.  (At least on the web; I could see e.g. emailing the user).

timabbott commented on templates/zerver/auth_subdomain.html:
```diff
@@ -0,0 +1,25 @@
+                        is meant to be an intermediate step in the authentication process
+                        and shouldn't be accessed manually. If you came here directly,
+                        you probably got the address wrong. If you got stuck here while trying
+                        to log in, this is most likely a server bug or misconfiguration.
```
Could we just serve a redirect to /accounts/go rather than making this page?  I have some design concerns about it (E.g. it shows as a 500 even though it isn't, etc.), and I think it might be cleaner to just not have it exist.

mateuszmandera replied:
Hmm, what do you mean by ` it shows as a 500`? Just the visuals of the page that opens due to reusing the template we normally use for 500 pages? I figured it fit visually for what this page is supposed to convey, but we could use a very basic template like for invalid confirmation links as well. The actual status code of the response was 200 - I changed this to return `400` now because it seems more appropriate.

We could just return /accounts/go, but my reasoning for having a special page was about the kind of circumstances a user could possibly bump into the page - it generally should only happen if something goes wrong in the authentication flow, and that'll usually indicate a bug. So it doesn't seem helpful to the user to show them an "Find your accounts" page when the actual problem has nothing to do with it. A scenario where a user opens `auth.zulipchat.com` should be rarer - because why would they do that, `auth.zulipchat.com` doesn't show up in any user-facing places.

timabbott replied:
Yeah, I think just showing the 500 error page template is the right model for something that's "Some sort of unexpected bug in the software", and avoids our having to design an additional page here.

timabbott commented on zerver/lib/name_restrictions.py:
```diff
@@ -1,7 +1,10 @@
 
 def is_reserved_subdomain(subdomain: str) -> bool:
+    if subdomain in settings.SOCIAL_AUTH_SUBDOMAIN:
+        return True
```
I think you want `if string_id in [settings.SOCIAL_AUTH_SUBDOMAIN]:` -- it's not a list.  This is doing substring comparisons.  Or at least if that's intentional, we should have a comment noting it.

mateuszmandera replied:
Yeah this should be `==`, I'm pretty sure this was a mistake

timabbott commented on tools/ci/success-http-headers.template.txt:
```diff
@@ -9,11 +9,21 @@ x-frame-options: DENY
+x-frame-options: DENY
+x-content-type-options: nosniff
+x-xss-protection: 1; mode=block
+
```
This should be squashed into the commit that changes the actual behavior to require it.

timabbott commented:
For the CodeQL error, I think you could reasonably just log the user ID rather than the email, and that'd be in line with how we try to log user details elsewhere.

timabbott commented:
This is looking good; sorry for the delay in trying to integrate it.  I posted a batch of comments that I think is reasonably complete, and also merged the first commit as 86c330b7520714e5e78ae7751d18fbc928096eb5.

mateuszmandera committed:
auth: Adjust behavior of /login/ on the root subdomain.

The core issue this is meant to resolve is that if a deployment has
its organizations on non-root subdomains, but doesn't set
ROOT_DOMAIN_LANDING_PAGE=True then the /login/ page uses its "normal
behavior" on the root subdomain even though there's nothing there -
which doesn't make sense and leads to exceptions being raised if
login+password is POSTed to it.

We should redirect to /accounts/go/ if there's no realm on the root
subdomain. A bit more subtle adjustments need to be made for the
dev_auth_enabled() case, in particular to not break
zulipdev.com:9991/devlogin/ as
a way to access the dev environment.

mateuszmandera commented:
@timabbott Ready for another round of review. Replied with some elaboration on the special auth subdomain page in https://github.com/zulip/zulip/pull/13924#discussion_r694124822

Regarding CodeQL and https://github.com/zulip/zulip/pull/13924#discussion_r676051716
> This isn't new in this PR, but I kinda feel like this "wrong subdomain" error category should be removed; it feels like we should just tell the user it was a failed login rather than leaking their credentials were correct but for the wrong domain. (At least on the web; I could see e.g. emailing the user).

CodeQL error and this comment are about the same block of code:
```
            if return_data.get("invalid_subdomain"):
                logging.warning(
                    "User %s attempted password login to wrong subdomain %s", username, subdomain
                )
                error_message = WRONG_SUBDOMAIN_ERROR.format(username=username)
                raise ValidationError(mark_safe(error_message))
```

I think we should get rid of the special error in favor of the generic one as you mentioned, and also get rid of the `logging.warning`? I don't know if someone trying to auth with the credentials on the wrong subdomain is interesting enough to log? And it only generates the CodeQL error

timabbott commented:
Logging the user ID rather than the username would fix the logging issue.

I found this class of logging statements to be useful when we had the caching incident a few months ago (though I think it was a slightly different version), so I think it's better to keep them (maybe at INFO level), just adjusted to only contain user IDs, not emails.

mateuszmandera committed:
auth: Don't log email when user tried authing to the wrong subdomain.

We want to avoid logging this kind of potentially sensitive information.
Instead, it's more useful to log ids of the matching accounts on
different subdomains.

mateuszmandera committed:
auth: Change the logger for authing to the wrong subdomain situations.

It's better to use zulip.auth logger here instead of the root logger.
Also the level is adjusted to INFO, since WARNING is excessive.

mateuszmandera committed:
auth: Don't leak info about accounts in other realms in the login form.

The "wrong subdomain" error leaks information that we shouldn't reveal.
We want to return a generic failure message instead.

mateuszmandera commented:
@timabbott Made changes to the logging and removed the user-facing "wrong subdomain" error. Is there anything remaining we want to do here?

timabbott commented:
https://github.com/zulip/zulip/pull/13924#discussion_r704800500 still needs a resolution -- can we just do a redirect rather than adding that page?

timabbott commented:
I wanted to merge the first half of this branch, which I feel relatively good about , so I did that as the series ending with f064e3ebac202e415de00b4a47bd439b0c0de954.  I think it's likely still worth trying to replace fb3864ea3cb2aaa27b10506cb4618f28a95c74d4 with a redirect for clarity.  

I have some concerns about the remaining commits (or just don't understand them), will post a few comments.

timabbott commented on zerver/views/auth.py:
```diff
@@ -756,7 +756,16 @@ def login_page(
+        # ROOT_DOMAIN_LANDING_PAGE is a setting for enabling special "landing page" behavior
+        # on the root subdomain, so we also want to redirect away from the login page if it's enabled.
+        and (settings.ROOT_DOMAIN_LANDING_PAGE or not Realm.objects.filter(string_id="").exists())
+    ):
```
I don't understand the set of conditionals here.  It might be better to reorder this logic so that the existing `dev_auth_enabled` conditional is just above this block, for example, to make it all easier to understand?

mateuszmandera commented:
@timabbott
> > I don't understand the set of conditionals here. It might be better to reorder this logic so that the existing `dev_auth_enabled` conditional is just above this block, for example, to make it all easier to understand?

I've extracted the overcomplicated if to a `should_redirect_away_from_login_page` function which can have a much clearer conditional structure, does that help?

> https://github.com/zulip/zulip/pull/13924#discussion_r704800500 still needs a resolution -- can we just do a redirect rather than adding that page?

I'm not sure what we should redirect to though? Is the idea still to redirect to `/accounts/go`?
